Refactor: Enhance HatsProposalCreationWhitelistV1 and Integrations

adamgall opened the pull request:
Depends on https://github.com/decentdao/decent-contracts/pull/245

Fixes [ENG-816](https://linear.app/decent-labs/issue/ENG-816/refactor-standardize-and-improve-hatsproposalcreationwhitelistv1)

**Summary:**

This PR refactors the `HatsProposalCreationWhitelistV1` contract and its related components for improved clarity, consistency, and Ownable pattern adherence. The changes include renaming functions and events, adjusting initialization, and updating all consuming contracts, interfaces, and tests accordingly.

**Key Changes:**

1.  **`HatsProposalCreationWhitelistV1.sol` Refinements:**
    *   **Initialization:** The `initialize` function now explicitly takes an `_owner` parameter and calls `__Ownable_init(_owner)` for proper ownership setup.
    *   **Error Renaming:** The error `InvalidHatsContract` has been renamed to `MissingHatsContract` for better semantic meaning.
    *   **Function Renaming & Visibility:**
        *   The internal `_whitelistHat` function has been made `public onlyOwner` and renamed to `whitelistHat`. The previous external `whitelistHat` function was removed to avoid redundancy.
        *   `removeHatFromWhitelist` has been renamed to `unwhitelistHat` for conciseness.
        *   The core proposer-checking function `isProposer` has been renamed to `isWearingWhitelistedHat` to more accurately describe its behavior (checking if an address wears any of the whitelisted hats).
    *   **Event Renaming:** The event `HatRemovedFromWhitelist` is now `HatUnwhitelisted`.
    *   Natspec comments have been updated throughout the contract to reflect these changes.

2.  **Updates to Consuming Strategy Contracts:**
    *   **`LinearERC20VotingWithHatsProposalCreationV1.sol`**
    *   **`LinearERC721VotingWithHatsProposalCreationV1.sol`**
        *   The `initialize` function in `HatsProposalCreationWhitelistV1` is now called with the `_owner` parameter.
        *   Unused imports (`Version`, `BaseStrategyV1`) have been removed.
        *   The overridden `isProposer` function now calls the renamed `isWearingWhitelistedHat` from the parent.

3.  **Interface `IHatsProposalCreationWhitelistV1.sol` Alignment:**
    *   All function names, event names, and error names have been updated to match the changes in the concrete `HatsProposalCreationWhitelistV1` contract.

4.  **Mock Contract `ConcreteHatsProposalCreationWhitelistV1.sol` Update:**
    *   The `initialize` function has been updated to accept and pass the `_owner` parameter to the parent's `initialize` and `__Ownable_init`.

5.  **Test Suite `HatsProposalCreationWhitelistV1.test.ts` Modifications:**
    *   Initialization calls in tests now correctly pass the `owner` address.
    *   Assertions for error names (`MissingHatsContract`) have been updated.
    *   All test logic and event checks now use the new function names (`whitelistHat`, `unwhitelistHat`, `isWearingWhitelistedHat`) and event name (`HatUnwhitelisted`).
    *   Type casting for UUPS upgradeability tests has been adjusted.

6.  **Helper Function `hatsProposerTests.ts` Adjustments:**
    *   The `HatsProposerTester` interface and its usage within the shared test logic have been updated to use `isWearingWhitelistedHat` instead of `isProposer`.

**Motivation:**

*   To improve the clarity and expressiveness of the `HatsProposalCreationWhitelistV1` contract's API.
*   To ensure correct and explicit ownership initialization following OpenZeppelin's `Ownable` pattern.
*   To maintain consistency across the contract, its interface, consuming contracts, and test suites.
*   To remove potential redundancies (e.g., separate internal and external `whitelistHat` functions).

mudrila commented on contracts/deployables/strategies/HatsProposalCreationWhitelistV1.sol:
```diff
@@ -29,39 +29,30 @@ abstract contract HatsProposalCreationWhitelistV1 is
-     * Adds a Hat to the whitelist for proposal creation.
-     * @param _hatId The ID of the Hat to whitelist
-     */
-    function whitelistHat(uint256 _hatId) external onlyOwner {
```
Wondering why we're removing this? If it's not necessary - why was it here in the first place?

adamgall replied:
I just changed the visibility from `external` to `public`, and then I actually did remove the `_whitelistHat internal` function.

The reason was that now that this contract calls `__Ownable_init` with the proper owner before setting up the whitelisted hats state, I can just call the public function (with the `onlyOwner` modifier on it) directly from the initializer.

Previously I was unable to do this due to ownership issues.

mudrila commented on contracts/deployables/strategies/HatsProposalCreationWhitelistV1.sol:
```diff
@@ -87,18 +78,17 @@ abstract contract HatsProposalCreationWhitelistV1 is
-     * is wearing any of them using the Hats Protocol.
      */
-    function isProposer(address _address) public view virtual returns (bool) {
+    function isWearingWhitelistedHat(
```
Nearly not getting why we removed `isProposer` - but I guess I need to look more closely into architectural shift in the versioned contracts. Just wanna make sure this `isProposer` removal is intentional

adamgall replied:
I think you found the answer to this below, but the reason was simply that the function didn't _need_ to be called `isProposer`, and I wanted to remove the purely linguistic link to "proposals" from this contract. This contract only deals with adding/removing/checking whitelisted hats and if addresses wear them. It shouldn't/doesn't care about anything downstream about how that information is used, and leaving "isProposer" in here crossed that line.

mudrila commented on contracts/deployables/strategies/LinearERC20VotingWithHatsProposalCreationV1.sol:
```diff
@@ -77,17 +76,10 @@ contract LinearERC20VotingWithHatsProposalCreationV1 is
-    {
-        return HatsProposalCreationWhitelistV1.isProposer(_address);
+    ) public view virtual override returns (bool) {
+        return isWearingWhitelistedHat(_address);
```
Aha, there you go

adamgall replied:
yup

mudrila commented on contracts/deployables/strategies/LinearERC721VotingWithHatsProposalCreationV1.sol:
```diff
@@ -80,17 +79,10 @@ contract LinearERC721VotingWithHatsProposalCreationV1 is
-    {
-        return HatsProposalCreationWhitelistV1.isProposer(_address);
+    ) public view virtual override returns (bool) {
+        return isWearingWhitelistedHat(_address);
```
Boom

adamgall replied:
yup

mudrila commented on test/helpers/hatsProposerTests.ts:
wen `foundry` / `forge`

mudrila reviewed (approved):
Gogogogo

adamgall committed:
Refactor: Enhance HatsProposalCreationWhitelistV1 and Integrations

**Summary:**

This PR refactors the `HatsProposalCreationWhitelistV1` contract and its related components for improved clarity, consistency, and Ownable pattern adherence. The changes include renaming functions and events, adjusting initialization, and updating all consuming contracts, interfaces, and tests accordingly.

**Key Changes:**

1.  **`HatsProposalCreationWhitelistV1.sol` Refinements:**
    *   **Initialization:** The `initialize` function now explicitly takes an `_owner` parameter and calls `__Ownable_init(_owner)` for proper ownership setup.
    *   **Error Renaming:** The error `InvalidHatsContract` has been renamed to `MissingHatsContract` for better semantic meaning.
    *   **Function Renaming & Visibility:**
        *   The internal `_whitelistHat` function has been made `public onlyOwner` and renamed to `whitelistHat`. The previous external `whitelistHat` function was removed to avoid redundancy.
        *   `removeHatFromWhitelist` has been renamed to `unwhitelistHat` for conciseness.
        *   The core proposer-checking function `isProposer` has been renamed to `isWearingWhitelistedHat` to more accurately describe its behavior (checking if an address wears any of the whitelisted hats).
    *   **Event Renaming:** The event `HatRemovedFromWhitelist` is now `HatUnwhitelisted`.
    *   Natspec comments have been updated throughout the contract to reflect these changes.

2.  **Updates to Consuming Strategy Contracts:**
    *   **`LinearERC20VotingWithHatsProposalCreationV1.sol`**
    *   **`LinearERC721VotingWithHatsProposalCreationV1.sol`**
        *   The `initialize` function in `HatsProposalCreationWhitelistV1` is now called with the `_owner` parameter.
        *   Unused imports (`Version`, `BaseStrategyV1`) have been removed.
        *   The overridden `isProposer` function now calls the renamed `isWearingWhitelistedHat` from the parent.

3.  **Interface `IHatsProposalCreationWhitelistV1.sol` Alignment:**
    *   All function names, event names, and error names have been updated to match the changes in the concrete `HatsProposalCreationWhitelistV1` contract.

4.  **Mock Contract `ConcreteHatsProposalCreationWhitelistV1.sol` Update:**
    *   The `initialize` function has been updated to accept and pass the `_owner` parameter to the parent's `initialize` and `__Ownable_init`.

5.  **Test Suite `HatsProposalCreationWhitelistV1.test.ts` Modifications:**
    *   Initialization calls in tests now correctly pass the `owner` address.
    *   Assertions for error names (`MissingHatsContract`) have been updated.
    *   All test logic and event checks now use the new function names (`whitelistHat`, `unwhitelistHat`, `isWearingWhitelistedHat`) and event name (`HatUnwhitelisted`).
    *   Type casting for UUPS upgradeability tests has been adjusted.

6.  **Helper Function `hatsProposerTests.ts` Adjustments:**
    *   The `HatsProposerTester` interface and its usage within the shared test logic have been updated to use `isWearingWhitelistedHat` instead of `isProposer`.

**Motivation:**

*   To improve the clarity and expressiveness of the `HatsProposalCreationWhitelistV1` contract's API.
*   To ensure correct and explicit ownership initialization following OpenZeppelin's `Ownable` pattern.
*   To maintain consistency across the contract, its interface, consuming contracts, and test suites.
*   To remove potential redundancies (e.g., separate internal and external `whitelistHat` functions).
